File: V3VCore/StereoConverter_Seperate/src/ViewInterpolation.py
import cv2
import numpy as np
from scipy.misc.pilutil import imfilter
from numpy import float64, float32
from scipy.sparse.construct import spdiags
from scipy.sparse.linalg.isolve.iterative import cg
from cv2 import BORDER_TRANSPARENT
from StereoConverter import average_before_remap, average_cg, average_prepare_cg, average_remap, average_clip, average_ss
import time
"""
RAW: Original Left image
RAW2: Left image resized by R
D: Depth Image (2D) resized by R
Thresh: Disparity range...set to 20
R: resize factor. Set to 0.25
Gx,Gy: See GetGxGy. Estimate ONLY ONCE for entire sequence and save in RAM (saves time)
Output SS: Side by Side output
"""
def ViewInterpolation(RAW, RAW2, D, Thresh, R, Gx, Gy, xx, yy, YY):
    
    # Warping
    lefts, rights, disparity = stereoWarpK_noMotion_singleSided(RAW2, RAW, D, Gx, Gy, Thresh, R, xx, yy, YY);
    
    aa = time.time()
    #Create SS Side by Side image
    vres, hres, u = lefts.shape
    SS = np.zeros((vres, hres*2, 3), dtype=np.float32)
    SS[:, 0:hres, :]  = lefts
    SS[:, hres: , :]  = rights
    bb = time.time()
    average_ss(bb-aa)
            
    return SS

"""
This function warps an image sequence (with corresponding flow and depth) to a stereo sequence
Assuming [h,w] are the video dimensions and K is the number of frames
# Input
imgs is of size [h,w,3,K] - video frames resized by R
imgs_RAW: video frames with original size
depths of size [h,w,1,K] - depth frames (grayscale) resized by R
flows of size [h,w,2,K] - optical flow results [????]
displevels - maximum amount of disparity between resulting stereo frames (in pixels) [????]
# Output:
lefts/rights - estimated stereo frames (same size as imgs)
anaglyphs - stereo frames converted to anaglyph [Where ????]
disparity - estimated disparity frames
"""
def stereoWarpK_noMotion_singleSided(imgs, imgs_RAW, depths, Gx, Gy, displevels, R, xx, yy, YY):
    ## Spatial smoothness parameters (higher => smoother)
    smoothCoeff_x = 20
    smoothCoeff_y = 4
    
    ## Prepare optimization
    
    h, w, u = imgs.shape # shape after resize
    K = 1
    N = h * w * K
    gr = np.mean(imgs, 2) # not 3 as it is zero based :3
    grs = cv2.GaussianBlur(gr, (5, 5), 1)
    
    # One heuristic for converting depth to disparity
    disparity0 = imnormalize(1/(1+imnormalize(depths)))*displevels - displevels/2;
    
    # Vertical and Horizontal Edges
    dx = cv2.filter2D(grs, -1, np.transpose(np.array([[-1, 1, 0]])))
    dy = cv2.filter2D(grs, -1, np.array([[-1, 1, 0]]))
    
    W = ( imnormalize(disparity0) + sigmoid(np.sqrt(np.power(dx, 2) + np.power(dy, 2)), 0.01, 500) ) / 2  
    
    aa = time.time()
###
    A = np.transpose(spdiags(np.transpose(W).flatten(), 0, N, N, "csc") + (smoothCoeff_x * Gx.transpose() * Gx) + (smoothCoeff_y * Gy.transpose() * Gy))
    bb = time.time()
    average_prepare_cg(bb-aa)
    
    b = np.transpose(W).flatten() * np.transpose(disparity0).flatten()
    
    # not the modified version
    # M = np.linalg.cholesky(A.todense())
    aa = time.time()
    [x, flag] = cg(A, b, np.transpose(disparity0).flatten(), 5e-1, 50)
    bb = time.time()
    average_cg(bb-aa)
    
    disparity = np.transpose(np.reshape(x, (w, h))) # remove (h, w, 1, K)

    warpright = -disparity
    
    lefts = imgs_RAW;
    
    warpright = cv2.resize(warpright, (imgs_RAW.shape[1], imgs_RAW.shape[0]), interpolation=cv2.INTER_LINEAR)
    
    rights = (clip(warpImage_v2((imgs_RAW), (warpright), R, xx, yy, YY)))
    
    #rights = (clip(warpImage_v2((imgs), (warpright), R, xx, yy, YY)))
    #rights = cv2.resize(rights, (imgs_RAW.shape[1], imgs_RAW.shape[0]), interpolation=cv2.INTER_LINEAR)
    
    return lefts, rights, disparity
    
def imnormalize(img, low_prc=None, up_prc=None):
    I = np.array(img)
    if type(I) != float32:
        I = float32(I)
    if abs(np.max(I)-np.min(I)) < 1e-8:
        I_norm = img
    else:
###
        if low_prc == None or up_prc == None:
            I_norm = (I - np.min(I)) / (np.max(I) - np.min(I))
        else:
            m = np.percentile(I, low_prc)
            M = np.percentile(I, up_prc)
            I_norm = (I-m)/(M-m);
    return I_norm
    
def sigmoid(x, center, scale):
    sx = 1. / (1 + np.exp(scale * (center-x)))
    return sx

# make all elements between bds[0] and bds[1]
def clip(v, bds=None):
    aa = time.time()
    if bds == None:
        bds = [0,1]
###
    v = np.array(v)
    v[v>bds[1]] = bds[1]
    v[v<bds[0]] = bds[0]
    bb = time.time()
    average_clip(bb-aa)
    return v
    
# function to warp images with different dimensions
def warpImage_v2(im, vx, R, xx, yy, YY):
    
    aa = time.time()

    height, width, nchannels = im.shape
    
    XX = xx + vx
    YY = YY
    # The brackets are very important :3
    mask = (XX<1) | (XX>width) | (YY<1) | (YY>height)
    #XX = np.minimum(np.maximum(XX, 1), width)
    bb = time.time()
    average_before_remap(bb-aa)
    
    aa = time.time()
    warpI2 = np.zeros((height, width, nchannels))
###
    for i in range(0, nchannels):
        foo = cv2.remap(im[:,:,i], XX.astype(float32)-1, YY.astype(float32)-1, interpolation = cv2.INTER_LINEAR, borderMode=BORDER_TRANSPARENT)
        foo[mask] = 0.6;
        warpI2[:, :, i] = foo
    bb = time.time()
    average_remap(bb-aa)
    
    # mask = 1 - mask
    
    return warpI2


    

File: V3VCore/StereoConverter_Seperate/MATLAB Code/ViewInterpolation.m
function SS = ViewInterpolation(RAW,D,Thresh,R,Gx,Gy,xx,yy,YY)

%%RAW: Original Left image
%%D: Depth Image (2D)
%%Thresh: Disparity range...set to 20
%%R: resize factor. Set to 0.25
%%Gx,Gy: See GetGxGy. Estimate ONLY ONCE for entire sequence and save in
%%RAM (saves time)

%%SS: Side by Side output

%%Make sure RAW and D is in the range of 0-1 (you can convert them usnig
%%im2double

I = imresize(RAW,R);
D = imresize(D,[size(I,1) size(I,2)]);

%%%Warping

[lefts, rights, disparity] = stereoWarpK_noMotion_singleSided(I, RAW, D, Gx,Gy, Thresh,R,xx,yy,YY);

%%%Creat Side by Side
[vres hres u] = size(lefts);
SS = zeros(vres,hres*2,3);
SS(:,1:hres,:) = lefts;
SS(:,hres+1:end,:) = rights;

imwrite(SS, '/Users/OmarEltobgy/Dropbox/WarpingTest1/SS1/3.png', 'png');
imwrite(SS, '/Users/OmarEltobgy/Dropbox/WarpingTest1/RGB/21.png', 'png');

end

function [lefts, rights, disparity] = stereoWarpK_noMotion_singleSided( imgs, imgs_RAW, depths, Gx, Gy, displevels,R,xx,yy,YY)
%STEREOWARP warps an image sequence (with corresponding flow and depth) to a stereo sequence
%Input:
%Assuming [h,w] are the video dimensions and K is the number of frames:
%imgs is of size [h,w,3,K] - video frames
%depths of size [h,w,1,K] - depth frames
%flows of size [h,w,2,K] - optical flow results
%displevels - maximum amount of disparity between resulting stereo frames (in pixels)
%Output:
%lefts/rights - estimated stereo frames (same size as imgs)
%anaglyphs - stereo frames converted to anaglyph
%disparity - estimated disparity frames

%% Spatial + temporal smoothness parameters (higher => smoother)
smoothCoeff_x = 20; %20
smoothCoeff_y = 4; %4 %1
%smoothCoeff_tem = 0; %0

%% Prepare optimization
%fprintf('Preparing optimization...');
%preparetime = tic;

[h,w,~,K] = size(imgs);
if( ~exist('displevels','var') )
    displevels = floor(sqrt(h*w)/30);
end
N = h*w*K;
gr = mean(imgs,3);
grs = imfilter(gr, fspecial('gaussian', [5,5],1));
%eps = 1e-4;


%One heuristic for converting depth to disparity
disparity0 = imnormalize(1./(1+imnormalize(depths))).*displevels - displevels/2;

%%disparity0 = imnormalize(1./(1+imnormalize(depths))).*displevels - 0*displevels/2;

dx = imfilter(grs, [-1 1 0], 'replicate', 'same');
dy = imfilter(grs, [-1 1 0]', 'replicate', 'same');
W = (imnormalize(disparity0) + 1.*sigmoid(sqrt(dx.^2+dy.^2),0.01,500))./2;

%A = spdiags(W(:),0,N,N) + smoothCoeff_x.*(Gx'*Gx) + smoothCoeff_y.*(Gy'*Gy) + smoothCoeff_tem.*(Gt'*wt*Gt);
t2 = smoothCoeff_x.*(Gx'*Gx);
A = spdiags(W(:),0,N,N) + smoothCoeff_x.*(Gx'*Gx) + smoothCoeff_y.*(Gy'*Gy);
b = W(:).*disparity0(:);


%fprintf('done! (%.5fs)\n', toc(preparetime));

%% Optimize
%fprintf('Optimizing warps...\n');
%opttime = tic;

%x = A\b;
%fprintf('\tConstructing preconditioner...\n');
M = ichol(A,struct('michol','on'));
%fprintf('\tOptimizing...\n\t');
%tic
[x flag] = pcg(A,b,5e-1,150,M,M',disparity0(:)); %1-e6
%toc

disparity = reshape(x,[h,w,1,K]);
%warpleft = disparity./2;
warpright = -disparity./1;

%tic
%lefts = clip(warpImage_v2(imgs_RAW, warpleft, zeros(h,w),R));
lefts = imgs_RAW;
%rights = clip(warpImage_v2(imgs_RAW, warpright, zeros(h/R,w/R),R));
%rights = gather(clip(warpImage_v2(gpuArray(imgs_RAW), gpuArray(warpright), gpuArray(zeros(h,w)),R)));
warpright = imresize(warpright,[size(imgs_RAW,1) size(imgs_RAW,2)]);
%rights = clip(warpImage_v2(imgs_RAW, warpright,R,xx,yy,YY));

rights = (clip(warpImage_v2((imgs_RAW), (warpright), R, xx, yy, YY)));

end

function [ I_norm ] = imnormalize( img, low_prc, up_prc )
I = img;
if ~isfloat(I)
    I = double(I);
end
if( abs(max(I(:))-min(I(:)))<1e-8 )
    I_norm = img;
else
    if(nargin==1)
        I_norm = (I - min(I(:))) / (max(I(:)) - min(I(:)));
    else
        m = prctile(I(:),low_prc);
        M = prctile(I(:),up_prc);
        I_norm = (I-m)/(M-m);
    end
end
end

function sx = sigmoid(x, center, scale)
sx = 1./(1+exp(scale.*(center-x)));
end

function cv = clip(v, bds)
if(nargin<2)
    bds = [0,1];
end
cv = v;
cv(v(:)<bds(1)) = bds(1);
cv(v(:)>bds(2)) = bds(2);
end


% function to warp images with different dimensions
function [warpI2,mask]=warpImage_v2(im,vx,R,xx,yy,YY)

S = 1/R;

[height,width,nchannels]=size(im);
% [xx,yy]=meshgrid(1:width,1:height);
% YY=yy;
% YY=min(max(YY,1),height);

%vx = imresize(vx,S);
XX=xx+vx;

mask=XX<1 | XX>width | YY<1 | YY>height;
XX=min(max(XX,1),width);
XX = (XX);

%tic
for i=1:nchannels
    %foo=interp2(xx,yy,im(:,:,i),XX,YY,'bicubic');
    foo=interp2(xx(1,:),yy(:, 1),im(:,:,i),XX,YY);
    %foo=qinterp2(xx,yy,im(:,:,i),XX,YY);
    foo(mask)=0.6;
    warpI2(:,:,i)=foo;
end
%toc

mask=1-mask;

end
